doc/source/dev/roadmap-detailed.rst

```diff
@@ -123,8 +123,26 @@ feature requests.
 
 constants
 `````````
-This module is basically done, low-maintenance and without open issues.
+This module needs only periodic updates to the numerical values.
 
+differentiate
+`````````````
+This module was added with the understanding that its scope would be limited.
+The goal is to provide basic differentiation of black-box functions, not
+replicate all features of existing differentiation tools. With that in mind,
+items for future work include:
+
+- Improve support for functions that accept additional arguments (e.g. add
+  ``*args`` to ``jacobian`` and ``hessian``). Note that this is not trivial
+  due to the way elements of the arrays are eliminated when their corresponding
+  calculations have converged.
+- Improve implementation of `scipy.differentiate.hessian`: rather than chaining
+  first-order differentiation, use a second-order differentiation stencil.
+- Consider the addition of an option to use relative step sizes.
+- Consider generalizing the approach to use "polynomial extrapolation"; i.e.,
```

Comment:
The functions were originally developed to meet internal needs, and we are releasing them publicly because there has been demand for a while. We hope for the maintenance burden to be low if we keep the scope limited.  Of course, it's always up for discussion whether other methods should be added, but I phrased it like this to discourage developing features just for the sake of completness. I don't think we should try to compete with more specialized numerical differentiation packages.

> complex step

Perhaps there is space for it if there are not other vectorized, array API compatible implementations out there. Should they be contributed to other packages instead? Crossref gh-7195.

> quadrature based approaches

Are these actually distinct mathematically (that is, are net weights on each function value different?), or do they differ in implementation details?

Reply:
Complex step differentiation for hessians will suffer from roundoff, it seems:  https://www.acsu.buffalo.edu/~johnc/complex_step08.pdf

For 1st derivatives, `approx_derivative` implements it, https://github.com/scipy/scipy/blob/main/scipy/optimize/_numdiff.py#L612, and making it array api compatible looks straightforward.

Reply:
> Are these distinct mathematically (that is, are net weights on each function value different?), or do they just differ in derivation or implementation details? For instance, I think I found that Richardson extrapolation was more numerically stable for high order than what I have, but I believe it would be equivalent (for a given stencil) with exact arithmetic. (IIUC, if the weights have to cancel out the low-order terms of the Taylor series, there's not really much flexibility if you want to get an n th order rule with n + 1 function evaluations.)

I agree the foundation of all methods is approximate the function (although it doesn't necessarily have to be with a polynomial eg. rational functions are better if there are poles) and then differentiate the approximation. However if you allow function evaluations on a disk in the complex plane then you can avoid the issues associated with standard finite differences, see J. N. Lyness and C. B. Moler, Numerical differentiation of analytic functions or NUMERICAL ALGORITHMS BASED ON ANALYTIC FUNCTION
 VALUES AT ROOTS OF UNITY,
 ANTHONY P. AUSTIN
, PETER KRAVANJA, AND LLOYD N. TREFETHEN. `mpmath`'s `diff` offers a method along these lines.
Note, I have no issue with the roadmap as it stands I just wanted to raise a discussion point:)

Reply:
Right, that sounds like a generalization of the complex step? I meant for real steps.